### PR TITLE
[workspace] Update PETSc to latest release v3.18.0

### LINKED ISSUE #18031: petsc needs update to latest
petsc needs to be updated to the latest release. `baij.patch` should be dropped, but another patch is also causing issues.

It might not hurt to update petscconf.h either.

### PR DESCRIPTION
**Pull request by xuchenhan-tri**
Closes #18031

### CODE REVIEW

**Comment by xuchenhan-tri** [comment]
@drake-jenkins-bot linux-focal-clang-bazel-experimental-undefined-behavior-sanitizer please.
@drake-jenkins-bot linux-focal-clang-bazel-experimental-valgrind-memcheck please.

**Review by xuchenhan-tri — COMMENTED** [review]
+@mwoehlke-kitware for feature review please.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038#-:-NDclpJX4Iu-w_6AzugS:box1nkv)* status: LGTM missing from assignee mwoehlke-kitware, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @mwoehlke-kitware)

File: tools/workspace/petsc/patches/inherit.patch [hunk 1]
```diff
@@ -1,28 +1,37 @@
...
 +  static atomic_int atomic_idcnt = 1;
++  return atomic_fetch_add(&atomic_idcnt, 1);
+ }
+
+ /*
+@@ -55,14 +56,6 @@ PetscErrorCode PetscHeaderCreate_Private(PetscObject h, PetscClassId classid, co
+
+   PetscCall(PetscCommDuplicate(comm, &h->comm, &h->tag));
+
+-  /* Increment and store current object creation index */
```
Comment (mwoehlke-kitware) [inline tools/workspace/petsc/patches/inherit.patch:27]: Why is this code no longer required?

**Comment by mwoehlke-kitware** [comment]
:lgtm:

**Comment by xuchenhan-tri** [comment]
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDdH_R86i-DZzrBFFqe:b-woymy8) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

My understanding is that the `h->cidx` here does object index book keeping for MPI. It's not thread safe and since we don't need MPI, the easiest fix is to get rid of it.
</blockquote></details>

@jwnimmer-tri could you take a look at this to verify my understanding is correct. (You have good experience mangling PETSc code.)

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri and @mwoehlke-kitware)

___
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDhmslDDMehWkPtOb3-:b83bbzh) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

@jwnimmer-tri could you take a look at this to verify my understanding is correct. (You have good experience mangling PETSc code.)
</blockquote></details>

If we grep the code for `CreationIdx`, we see only a few uses of it:

- In `pinit.c`, there is an alloc/free pair. It's only ever created, never read from or used for anything.
- In `tagm.c`, there is an alloc/free pair. It's only ever created, never read from or used for anything.
- In `inheric.c`, there is this patch where we remove the `++` operation on it, which we are removing.

Given that nobody ever reads from this value, I agree that incrementing it here seems to be dead code that we can remove.

The only question is whether _in the future_ petsc would start to depend on this value.

In case there is a simple fix other than deleting the code, it might be worth doing that to be more future-proof.

Basically if instead of deleting `h->cidx = (*cidx)++;` you replace it with `atomic_fetch_add`, does that placate our memory race checkers?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri and @mwoehlke-kitware)

___
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDhoZEF9FpiWk7bflJI:b-6qtali) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If we grep the code for `CreationIdx`, we see only a few uses of it:

- In `pinit.c`, there is an alloc/free pair. It's only ever created, never read from or used for anything.
- In `tagm.c`, there is an alloc/free pair. It's only ever created, never read from or used for anything.
- In `inheric.c`, there is this patch where we remove the `++` operation on it, which we are removing.

Given that nobody ever reads from this value, I agree that incrementing it here seems to be dead code that we can remove.

The only question is whether _in the future_ petsc would start to depend on this value.

In case there is a simple fix other than deleting the code, it might be worth doing that to be more future-proof.

Basically if instead of deleting `h->cidx = (*cidx)++;` you replace it with `atomic_fetch_add`, does that placate our memory race checkers?
</blockquote></details>

Ah, I see now my grepping maybe wasn't enough.  I didn't check whether `h->cidx` was used for anything, and it's a bit difficult to know.  I see a use in `garbage.c` at least.

I'm curious to know whether we can just switch to atomic instead of `++`.

**Review by xuchenhan-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri and @mwoehlke-kitware)

___
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDiKlM90EnmMfa9G5EB:b-t4gzr2) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ah, I see now my grepping maybe wasn't enough.  I didn't check whether `h->cidx` was used for anything, and it's a bit difficult to know.  I see a use in `garbage.c` at least.

I'm curious to know whether we can just switch to atomic instead of `++`.
</blockquote></details>

I tried something along that line e.g.
```
   /* Increment and store current object creation index */
   PetscCallMPI(MPI_Comm_get_attr(h->comm, Petsc_CreationIdx_keyval, &get_tmp, &flg));
   if (flg) {
-    cidx    = (PetscInt64 *)get_tmp;
-    h->cidx = (*cidx)++;
-    PetscCallMPI(MPI_Comm_set_attr(h->comm, Petsc_CreationIdx_keyval, cidx));
+    atomic_store(&atomic_cidx, *(int*)get_tmp);
+    atomic_fetch_add(&atomic_cidx, 1);
+    PetscCallMPI(MPI_Comm_set_attr(h->comm, Petsc_CreationIdx_keyval, &atomic_cidx));
   } else SETERRQ(h->comm, PETSC_ERR_ARG_CORRUPT, "MPI_Comm does not have an object creation index");
```

but that didn't work. The race condition is not eliminated by this -- I think we need the entire operation of "`MPI_Comm_get_attr` + incrementing by 1" to be atomic, but I'm not sure how to do that.

**Comment by jwnimmer-tri** [comment]
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDiLvfy118veF4C5wkk:bea9xqv) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

I tried something along that line e.g.
```
   /* Increment and store current object creation index */
   PetscCallMPI(MPI_Comm_get_attr(h->comm, Petsc_CreationIdx_keyval, &get_tmp, &flg));
   if (flg) {
-    cidx    = (PetscInt64 *)get_tmp;
-    h->cidx = (*cidx)++;
-    PetscCallMPI(MPI_Comm_set_attr(h->comm, Petsc_CreationIdx_keyval, cidx));
+    atomic_store(&atomic_cidx, *(int*)get_tmp);
+    atomic_fetch_add(&atomic_cidx, 1);
+    PetscCallMPI(MPI_Comm_set_attr(h->comm, Petsc_CreationIdx_keyval, &atomic_cidx));
   } else SETERRQ(h->comm, PETSC_ERR_ARG_CORRUPT, "MPI_Comm does not have an object creation index");
```

but that didn't work. The race condition is not eliminated by this -- I think we need the entire operation of "`MPI_Comm_get_attr` + incrementing by 1" to be atomic, but I'm not sure how to do that.
</blockquote></details>

Could you provide the relevant (and minimal) `bazel test` command for me to repro?  I'd like to poke around.

**Comment by xuchenhan-tri** [comment]
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDiPWYS6k6MxHOiBaFa:b6xeyu1) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Could you provide the relevant (and minimal) `bazel test` command for me to repro?  I'd like to poke around.
</blockquote></details>

I simply did `bazel test ...` to cover all bases. The specific test that's failing is (unsurprisingly) `petsc_symmetric_block_sparse_matrix_test`. So you can do `bazel test //multibody/fem:petsc_symmetric_block_sparse_matrix_test`.

**Comment by jwnimmer-tri** [comment]
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDiSsNC3IA7cDBD-1rR:b-vscf61) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

I simply did `bazel test ...` to cover all bases. The specific test that's failing is (unsurprisingly) `petsc_symmetric_block_sparse_matrix_test`. So you can do `bazel test //multibody/fem:petsc_symmetric_block_sparse_matrix_test`.
</blockquote></details>

Which compiler and/or bazel config flags though?

**Comment by xuchenhan-tri** [comment]
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDiWfdR9GEStV5EVsyG:b-6jaoec) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Which compiler and/or bazel config flags though?
</blockquote></details>

Default one should do. I ran both gcc and clang and both release and debug build. All give the same issue.
You can do `bazel  test --config=clang --compilation_mode=dbg //multibody/fem:petsc_symmetric_block_sparse_matrix_test` for example.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri and @mwoehlke-kitware)

___
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDidZ_H8uoGy8_k6pSo:b-ard4ez) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Default one should do. I ran both gcc and clang and both release and debug build. All give the same issue.
You can do `bazel  test --config=clang --compilation_mode=dbg //multibody/fem:petsc_symmetric_block_sparse_matrix_test` for example.
</blockquote></details>

Okay.  I think the best way to prune away this unused feature would be to add a (different) patch file that removes the `Petsc_CreationIdx_keyval` declaration from the header, and then purges all code that touches it.

That way, in case petsc ever adds anymore code that tries to use the CreationIdx down the foal, we'll fail-fast.

**Review by xuchenhan-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform) (waiting on @jwnimmer-tri, @mwoehlke-kitware, and @xuchenhan-tri)

___
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDixGmsEkaWn3FluR16:b1urkpi) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Okay.  I think the best way to prune away this unused feature would be to add a (different) patch file that removes the `Petsc_CreationIdx_keyval` declaration from the header, and then purges all code that touches it.

That way, in case petsc ever adds anymore code that tries to use the CreationIdx down the foal, we'll fail-fast.
</blockquote></details>

Working. SGTM!

**Comment by xuchenhan-tri** [comment]
@drake-jenkins-bot linux-focal-clang-bazel-experimental-undefined-behavior-sanitizer please.
@drake-jenkins-bot linux-focal-clang-bazel-experimental-valgrind-memcheck please.

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: platform.

I noticed they've repaved their code with `clang-format`, thus so much diff churn here.  Sigh.

> Reviewed 8 of 11 files at r1, 1 of 1 files at r2, 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038#-:-NDjHXaeBwnyoQQdEJKb:b-y18h5b)* status: 2 unresolved discussions (waiting on @xuchenhan-tri)

___
*[`tools/workspace/petsc/patches/petscimpl.patch` line 33 at r4](https://reviewable.io/reviews/robotlocomotion/drake/18038#-NDjHtxL8d9mAiJBBMBF:-NDjHtxL8d9mAiJBBMBG:b-a4xogv) ([raw file](https://github.com/robotlocomotion/drake/blob/c2d569cb83f0c8e1b20c51251aac6c995ffb5543/tools/workspace/petsc/patches/petscimpl.patch#L33)):*
> ```Diff
> 
>    #else
> ~        
> ```

nit I'm not entirely sure what reviewable is showing me on this line.  Please double-check that this patch file is as you intend.

**Review by xuchenhan-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: 1 unresolved discussion (waiting on @jwnimmer-tri and @mwoehlke-kitware)

___
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDjIoolDiGeOE0CkppE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Working. SGTM!
</blockquote></details>

Done.
___
*[`tools/workspace/petsc/patches/petscimpl.patch` line 33 at r4](https://reviewable.io/reviews/robotlocomotion/drake/18038#-NDjHtxL8d9mAiJBBMBF:-NDjOnun4wQC_f5V8NHl:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/c2d569cb83f0c8e1b20c51251aac6c995ffb5543/tools/workspace/petsc/patches/petscimpl.patch#L33)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit I'm not entirely sure what reviewable is showing me on this line.  Please double-check that this patch file is as you intend.
</blockquote></details>

Done

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: 1 unresolved discussion (waiting on @mwoehlke-kitware)

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot mac-big-sur-clang-bazel-experimental-release please

**Review by xuchenhan-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: 1 unresolved discussion (waiting on @mwoehlke-kitware)

___
*[`tools/workspace/petsc/patches/inherit.patch` line 27 at r2](https://reviewable.io/reviews/robotlocomotion/drake/18038#-ND_IvD7eGK2DnXb4P8G-r2-27:-NDjbZQ022l_FAZ5P4la:boikpeo) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfd1505981e3ba9d0e1be595c700ad7cffbc8c8/tools/workspace/petsc/patches/inherit.patch#L27)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Done.
</blockquote></details>

@mwoehlke-kitware this PR is ready to merge modulo this blocking comment. (I'm guessing this is unintentional but want to confirm.)

**Comment by mwoehlke-kitware** [comment]
:lgtm:

Reviewable is not behaving for me; I don't have a way to unblock anything it thinks is blocked.

**Review by jwnimmer-tri — COMMENTED** [review]
> Dismissed @mwoehlke-kitware from a discussion.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),mwoehlke-kitware (waiting on @xuchenhan-tri)

**Review by jwnimmer-tri — COMMENTED** [review]
> Dismissed @mwoehlke-kitware from a discussion.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/18038)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),mwoehlke-kitware (waiting on @xuchenhan-tri)

**Comment by jwnimmer-tri** [comment]
> LGTM

Thanks!  I've closed the discussion thread out to unblock.

> Reviewable is not behaving for me; I don't have a way to unblock anything it thinks is blocked.

In case this is still occurring, please open a support request to get it fixed.  We need Reviewable to be working for you.

As I understand it the support channels are:
- https://github.com/Reviewable/Reviewable/issues
- https://gitter.im/Reviewable/Reviewable
